### ISSUE TITLE
Fully mutating linear solver

### ISSUE THREAD

**Issue by gdalle** [gdalle/ImplicitDifferentiation.jl#48]
At the moment, AbstractDifferentiation does not have a fully mutating API, and benchmarks show allocations are non-negligible when multiplying by A and B. Opening this to track https://github.com/JuliaDiff/AbstractDifferentiation.jl/issues/14 and see what we can do

**Comment by mohamed82008** [comment]
The problem is that most AD packages don't actually support mutation.

**Comment by gdalle** [comment]
Yeah I know, it would be useless for ChainRules, but not for ForwardDiff?

**Comment by gdalle** [comment]
Trying to solve this with https://github.com/gdalle/DifferentiationInterface.jl

**Comment by gdalle** [comment]
This is now possible with DifferentiationInterface but

- it won't work with every outer backend
- it requires a mutating version of the conditions `conditions!(c, x, y)`

I hope to make it part of v0.7
